Annotate MediaStreamEnabled endpoints with feature flags

darryl-apple opened the pull request:
#### 1a3d5bb932c2c998b9d4e67b71ae055fae7dd506
<pre>
Annotate MediaStreamEnabled endpoints with feature flags
<a href="https://bugs.webkit.org/show_bug.cgi?id=283796">https://bugs.webkit.org/show_bug.cgi?id=283796</a>
<a href="https://rdar.apple.com/140655529">rdar://140655529</a>

Reviewed by Sihui Liu.

Annotate MediaStreamEnabled endpoints with feature flags

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.cpp:
(WebKit::GPUConnectionToWebProcess::updateSharedPreferencesForWebProcess):
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.messages.in:
* Source/WebKit/GPUProcess/webrtc/RemoteSampleBufferDisplayLayer.cpp:
(WebKit::RemoteSampleBufferDisplayLayer::sharedPreferencesForWebProcess const):
* Source/WebKit/GPUProcess/webrtc/RemoteSampleBufferDisplayLayer.h:
* Source/WebKit/GPUProcess/webrtc/RemoteSampleBufferDisplayLayer.messages.in:
* Source/WebKit/GPUProcess/webrtc/RemoteSampleBufferDisplayLayerManager.cpp:
(WebKit::RemoteSampleBufferDisplayLayerManager::RemoteSampleBufferDisplayLayerManager):
(WebKit::RemoteSampleBufferDisplayLayerManager::updateSharedPreferencesForWebProcess):
* Source/WebKit/GPUProcess/webrtc/RemoteSampleBufferDisplayLayerManager.h:
* Source/WebKit/GPUProcess/webrtc/RemoteSampleBufferDisplayLayerManager.messages.in:
* Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.cpp:
(WebKit::UserMediaCaptureManagerProxy::sharedPreferencesForWebProcess const):
* Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.h:
* Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.messages.in:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:

Canonical link: <a href="https://commits.webkit.org/287566@main">https://commits.webkit.org/287566@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e1275fec8821d1907fbb3eb7c746dfbb0d5e6b40

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/80066 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/59063 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/33468 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/84579 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/31036 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/68127 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/7360 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/62590 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/20414 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/83134 "Passed tests") | [⏳ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/iOS-17-Simulator-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/72922 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/42898 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/49995 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/29500 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/71118 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/27575 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/86008 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/7281 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/5146 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/70863 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/7458 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/68760 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/70100 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17463 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/14103 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/13044 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/7245 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/12773 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/7087 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/10605 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/8893 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

szewai commented on Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.messages.in:
```diff
@@ -23,7 +23,8 @@
 #if ENABLE(MEDIA_STREAM)
 
-messages -> UserMediaCaptureManagerProxy {
+    [EnabledBy=MediaPlaybackEnabled]
```
Extra blank spaces

szewai commented on Source/WebKit/UIProcess/WebProcessProxy.cpp:
```diff
@@ -312,6 +312,11 @@ class UIProxyForCapture final : public UserMediaCaptureManagerProxy::ConnectionP
 
+    std::optional<SharedPreferencesForWebProcess> sharedPreferencesForWebProcess() const
+    {
+        return m_process.get().sharedPreferencesForWebProcess();
```
`m_process->`?

szewai commented on Source/WebKit/GPUProcess/webrtc/RemoteSampleBufferDisplayLayerManager.h:
```diff
@@ -54,9 +55,9 @@ class RemoteSampleBufferDisplayLayer;
     WTF_MAKE_TZONE_ALLOCATED(RemoteSampleBufferDisplayLayerManager);
 public:
-    static Ref<RemoteSampleBufferDisplayLayerManager> create(GPUConnectionToWebProcess& connection)
+    static Ref<RemoteSampleBufferDisplayLayerManager> create(GPUConnectionToWebProcess& connection,  SharedPreferencesForWebProcess& sharedPreferencesForWebProcess)
```
Extra space before `SharedPreferencesForWebProcess`

szewai commented on Source/WebKit/GPUProcess/webrtc/RemoteSampleBufferDisplayLayerManager.cpp:
```diff
@@ -132,6 +136,13 @@ void RemoteSampleBufferDisplayLayerManager::updateSampleBufferDisplayLayerBounds
 
+void RemoteSampleBufferDisplayLayerManager::updateSharedPreferencesForWebProcess(SharedPreferencesForWebProcess sharedPreferencesForWebProcess)
+{
+    protectedQueue()->dispatch([this, sharedPreferencesForWebProcess = WTFMove(sharedPreferencesForWebProcess)] {
```
Need `protectedThis = Ref { *this }` in lambda

szewai commented on Source/WebKit/GPUProcess/webrtc/RemoteSampleBufferDisplayLayerManager.cpp:
```diff
@@ -41,11 +41,15 @@ namespace WebKit {
+    , m_sharedPreferencesForWebProcess(sharedPreferencesForWebProcess)
     , m_queue(gpuConnectionToWebProcess.protectedGPUProcess()->videoMediaStreamTrackRendererQueue())
 {
+    protectedQueue()->dispatch([this, sharedPreferencesForWebProcess] {
```
Need `protectedThis = Ref { *this }` in lambda

szewai commented on Source/WebKit/GPUProcess/webrtc/RemoteSampleBufferDisplayLayer.cpp:
```diff
@@ -167,6 +168,15 @@ void RemoteSampleBufferDisplayLayer::setShouldMaintainAspectRatio(bool shouldMai
 
+std::optional<SharedPreferencesForWebProcess> RemoteSampleBufferDisplayLayer::sharedPreferencesForWebProcess() const
+{
+    assertIsCurrent(m_consumeThread);
```
This assertion fails on bots.
Note `RemoteSampleBufferDisplayLayer` is created on main thread, but message handlers are invoked on `WorkQueue` of `RemoteSampleBufferDisplayLayerManager` (see `RemoteSampleBufferDisplayLayerManager::dispatchMessage`); so m_consumerThread is main and current thread is not.
We might need to make `RemoteSampleBufferDisplayLayer` gets the `preferences` from its manager (whose `m_sharedPreferencesForWebProcess` can be accessed from `WorkQueue`),  by making `RemoteSampleBufferDisplayLayer` have a `ThreadSafeWeakPtr< RemoteSampleBufferDisplayLayerManager >` member.

szewai reviewed (commented):
The [mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/4270) crashes seem relevant

szewai commented on Source/WebKit/GPUProcess/webrtc/RemoteSampleBufferDisplayLayer.cpp:
```diff
@@ -167,6 +169,12 @@ void RemoteSampleBufferDisplayLayer::setShouldMaintainAspectRatio(bool shouldMai
 
+std::optional<SharedPreferencesForWebProcess> RemoteSampleBufferDisplayLayer::sharedPreferencesForWebProcess() const
+{
+    auto remoteSampleBufferDisplayLayerManager = m_remoteSampleBufferDisplayLayerManager.get();
```
Technically manager could be null because `m_remoteSampleBufferDisplayLayerManager` is a weak pointer, so we should null check:
```
if (RefPtr manager = m_remoteSampleBufferDisplayLayerManager.get())
    return manager->sharedPreferencesForWebProcess();

return std::nullopt;
```

szewai reviewed (approved):
r=me if bots are happy

darryl-apple committed:
Annotate MediaStreamEnabled endpoints with feature flags
https://bugs.webkit.org/show_bug.cgi?id=283796
rdar://140655529

Reviewed by Sihui Liu.

Annotate MediaStreamEnabled endpoints with feature flags

* Source/WTF/Scripts/Preferences/UnifiedWebPreferences.yaml:
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.cpp:
(WebKit::GPUConnectionToWebProcess::updateSharedPreferencesForWebProcess):
* Source/WebKit/GPUProcess/GPUConnectionToWebProcess.messages.in:
* Source/WebKit/GPUProcess/webrtc/RemoteSampleBufferDisplayLayer.cpp:
(WebKit::RemoteSampleBufferDisplayLayer::sharedPreferencesForWebProcess const):
* Source/WebKit/GPUProcess/webrtc/RemoteSampleBufferDisplayLayer.h:
* Source/WebKit/GPUProcess/webrtc/RemoteSampleBufferDisplayLayer.messages.in:
* Source/WebKit/GPUProcess/webrtc/RemoteSampleBufferDisplayLayerManager.cpp:
(WebKit::RemoteSampleBufferDisplayLayerManager::RemoteSampleBufferDisplayLayerManager):
(WebKit::RemoteSampleBufferDisplayLayerManager::updateSharedPreferencesForWebProcess):
* Source/WebKit/GPUProcess/webrtc/RemoteSampleBufferDisplayLayerManager.h:
* Source/WebKit/GPUProcess/webrtc/RemoteSampleBufferDisplayLayerManager.messages.in:
* Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.cpp:
(WebKit::UserMediaCaptureManagerProxy::sharedPreferencesForWebProcess const):
* Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.h:
* Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.messages.in:
* Source/WebKit/UIProcess/WebPageProxy.messages.in:

Canonical link: https://commits.webkit.org/287566@main
